packages/core/src/deployment.ts

```diff
@@ -94,58 +101,113 @@ async function validateCached<T extends Deployment>(
         throw e;
       }
     }
+  } else {
+    return undefined;
   }
-  return cached;
 }
 
-async function validateStoredDeployment<T extends Deployment>(
+async function validateStoredDeployment<T extends Deployment & DeploymentId>(
   stored: T,
   provider: EthereumProvider,
   type?: string,
   opts?: DeployOpts,
   merge?: boolean,
-) {
-  const { txHash } = stored;
+  getDeploymentResponse?: (deploymentId: string, allowUndefined: boolean) => Promise<DeploymentResponse | undefined>,
+): Promise<T> {
+  const { txHash, deploymentId } = stored;
+  let deployment = stored;
+
   if (txHash !== undefined) {
     // If there is a deployment with txHash stored, we look its transaction up. If the
     // transaction is found, the deployment is reused.
     debug('found previous deployment', txHash);
     const tx = await getTransactionByHash(provider, txHash);
+    let foundDeployment = undefined;
     if (tx !== null) {
-      debug('resuming previous deployment', txHash);
+      foundDeployment = txHash;
+    } else if (deploymentId !== undefined && getDeploymentResponse !== undefined) {
+      // If the transaction is not found, try checking the deployment id since the transaction may have been replaced
+      const response = await getDeploymentResponse(deploymentId, true);
+      if (response !== undefined) {
+        foundDeployment = deploymentId;
+
+        // update the stored tx hash
+        deployment = { ...stored, txHash: response.txHash };
+        debug('updating previous deployment to tx hash ', response.txHash);
+      }
+    }
+
+    if (foundDeployment) {
+      debug('resuming previous deployment', foundDeployment);
       if (merge) {
         // If merging, wait for the existing deployment to be mined
-        waitAndValidateDeployment(provider, stored, type, opts);
+        waitAndValidateDeployment(provider, deployment, type, opts, getDeploymentResponse);
       }
     } else {
       // If the transaction is not found we throw an error, except if we're in
       // a development network then we simply silently redeploy.
       // This error should be caught by the caller to determine if we're in a dev network.
-      throw new InvalidDeployment(stored);
+      throw new InvalidDeployment(deployment);
     }
   } else {
-    const existingBytecode = await getCode(provider, stored.address);
+    const existingBytecode = await getCode(provider, deployment.address);
     if (isEmpty(existingBytecode)) {
-      throw new InvalidDeployment(stored);
+      throw new InvalidDeployment(deployment);
     }
   }
+  return deployment;
+}
+
+export interface DeploymentResponse {
```

Comment:
"Deployment response" doesn't seem like the right concept to me... I think it should be either `DeploymentStatus` or something like `RemoteDeployment` or even `PlatformDeployment`. This would apply as well to the function names.

I'm thinking we should go with the latter and use `remoteDeploymentId` in the manifest. `deploymentId` is very generic.